### DISCUSSION TITLE
Audio support

### DISCUSSION THREAD

**Discussion by ku1ik** [asciinema/discussions#96] [Ideas]
Possibly uploaded by user and played through html5 audio tag.

**Comment by icholy** [comment]
I don't think this is a good idea. asciiio should do one thing, do it simply, and do it well. If people want audio then asciiio is beyond the scope of what they need. Imo audio is just adding adding unnecessary complication.

**Comment by ku1ik** [comment]
Thank you for stating your opinion. In my opinion it will complicate code, I agree with you.  How much code it will require depends on the way it would be implemented. There are several ways this can be achieved, I'm not convinced by any of them yet.

I'm leaving this issue open to have a place for discussion about this feature.

**Comment by tjeden** [comment]
+1 for audio support. I would use ascii.io to create screencasts, but without audio it's useless for me.

**Comment by ku1ik** [comment]
There are 2 options for this:
1. allow user to upload any mp3/ogg with voiceover to his already recorded asciicast
2. record microphone during recording of asciicasts, encode to mp3/ogg, send mp3 together with text data

Both have upsides and downsides.

Option 1: Upside: user can record voiceover any number of times, remove noise etc. Downside: it's too much hassle for someone who just wants to do some quick recording, who doesn't want to deal with this additional "audio work".

Option 2: Upside: recording simplicity stays the same for the user. Downsides: it limits options of editing/preparing audio. Also it adds dependencies on cross-platform audio recording lib (for example pyaudio) and on mp3/ogg encoding tool.

**Comment by losingkeys** [comment]
This may warren its own issue/discussion, but what about integrating with [popcorn.js](https://popcorn.webmaker.org/)?  Also, the fact that long pauses are zipped to 5 seconds may become an issue if we want people talking over their presentation.  Using popcorn.js we might allow the to record the audio after... and could even provide an option for if they wanted the time zipped up or not.

**Comment by wxianfeng** [comment]
i want it support audio

**Comment by asmcos** [comment]
I had add mp3 support.
And I will support picture next time.

**Comment by mattboehm** [comment]
I would love audio support so I could use it for screencasts as well. I think it should not be the responsibility of the client, as that sounds too complicated to add.

Uploading mp3/ogg (or even just providing a link to an audio file hosted elsewhere so you don't have to host it) would definitely work.

If you don't want to add all that to your codebase, at the very least it would be nice if the embedded player exposed hooks so I could sync it with an audio player myself.

**Comment by nathany** [comment]
Audio support would be pretty cool. Most importantly, I think:
- Hooks in the player to sync starting audio and the cast. This would make it possible the play audio on a blog even if not directly supported on asciinema.org.
- A granular way to speed up portions of the cast to synchronize with audio, and reduce long pauses (or create pauses/stretch out the cast as needed).

My vote would be for option 1, upload audio later. This more closely matches how screencasts are made and doesn't make the recorder more complicated. Those that don't want to be bothered with audio could just add a textual transcript.

**Comment by ku1ik** [comment]
I like what @mattboehm and @nathany suggest. Let's go this way.

I have added "hooks" to the TODO list of the player (https://github.com/asciinema/asciinema-player) and will prioritize this work.

**Comment by kordless** [comment]
I just came here looking for some sort of sound syncing ability for the product (which I love BTW).  Vimeo has JS hooks to enable actions to take place external to the player.  I like the simplicity of the player, and it should stay true to that simplicity, but if the users want to do something fancy, the player should be able to output where it is at, what it is doing and whether it should be playing or not.  These features enable far more functionality than sound alone!

**Comment by rharriso** [comment]
@sickill, after adding hooks to the player, would the next step be integrating with some audio hosting service, like soundcloud?

Or, do you think hosting the audio through asciinema.org would be best?

**Comment by ku1ik** [comment]
@rharriso so the first step is to add hooks to the player ("on-start", "on-pause" etc), so you can control the audio playback yourself. For example you'll be able to add html5 audio tag and play/pause the audio file in sync with terminal playback.

Once the player has this, I'll be able to use it on asciinema.org. As for hosting the files, I don't plan to host them. asciinema.org is a free service, and while hosting the recordings doesn't cost me much, audio files would make it more expensive. So my current idea for this is to just let people specify audio file URL which will be used directly by audio tag. If soundcloud provides direct audio file URLs then you'll be able to use them.

**Comment by rharriso** [comment]
@sickill, Is this being worked on already? Or is it an open item?

**Comment by ku1ik** [comment]
@rharriso atm I'm (slowly) working on the new version of the player here: https://github.com/asciinema/asciinema-player/tree/next I expect to finish this in the next couple of weeks, then I'll be working on the hooks.

**Comment by rharriso** [comment]
I'm in the process of moving, so I don't have any free time to contribute. But after I get where I'm going, I'll check in.

**Comment by rharriso** [comment]
@sickill, I'm settling in after my move, and have a little more time to work on side stuff. How are you doing on this? Do you want some help implementing it?

**Comment by ku1ik** [comment]
@rharriso let's move the discussion about hooks to https://github.com/asciinema/asciinema-player/issues/12, as it needs to be done before we can add audio support to the site.

**Comment by weitzman** [comment]
As an update, https://github.com/asciinema/asciinema-player/issues/12 added the js events needed for for synching audio. So, we're not there yet but its getting close.

**Comment by anarcat** [comment]
@weitzman awesome! (and hi - fancy seeing you here :)) - does that mean there is a way to setup an audio feed next to asciinema *now*?

i mean i'm fine with recording audio using a separate tool and slapping it in there somehow, i just don't feel it's even *possible* with the current implementation...

**Comment by weitzman** [comment]
Hi. @anarcat. Fancy indeed.

It looks *possible*, but there isn't an example anywhere that I've found. The starting point would be the JS docs at https://github.com/asciinema/asciinema-player#controlling-the-player-programmatically

**Comment by anarcat** [comment]
riiight... so i guess it would probably be simpler to just code a audio player straight into the normal asciinema right here... otherwise you end up replicating the whole thing.

in other words, it's probably easier to pass *in* asciinema an audio file than try to extract the asciinema player *out* of asciinema and implement a separate audio player, considering the latter is basically standard HTML5...

**Comment by pfalcon** [comment]
https://github.com/dhobsd/castty is an asciinema-compatible recorder and player with audio support. Demo: https://9vx.org/%7Edho/term/index.html

**Comment by F1LT3R** [comment]
@weitzman can you manually set the position/time of the player with javascript'?

**Comment by Bikraji** [comment]
Guys, what is the status of this issue?
Are you going to support this feature, any road map?

**Comment by data-harmonization** [comment]
> https://github.com/dhobsd/castty is an asciinema-compatible recorder and player with audio support. Demo: https://9vx.org/%7Edho/term/index.html

For people looking into this alternative, my own findings are that Casttly is less advanced than asciinema with respect to the recording options. Also, I was not able to play a Casttly-cast with the Asciinema player (while the webpage says it is valid v1-format, so probably, I did something wrong). Last but not least, my audio recording on MacOS BigSur is unfortunately not playing at all. 

Bottom line, Casttly doesn't seem an alternative for folks like me who like to annotate their recordings with audio.

If I'm not mistaken, this is the only asciinema-issue related to audio/voice-support that is still open. I would love to hear if there are any updates or users that managed to add audio to their recordings

**Comment by cipy** [comment]
while indeed is best to keep asciinema simple and fast,

how about providing some placeholder code so that when we play the recording on a webpage, we'd get an audio track started at the same time. recording audio should be outside the scope of asciinema, as semi-pro's would like to edit, cut, refine, adjust the sound levels before going live with the track.

thanks 💯

**Comment by ku1ik** [comment]
FYI asciinema player supports synced audio since version 3.11.

https://github.com/asciinema/asciinema-player/releases/tag/v3.11.0

You can use it on asciinema.org too - in recording settings there's "Audio URL" option.

**Reply by balupton** [reply]
just needs the ability to record the audio too when recording the asciinema cast
